data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisCheckpointerTracker.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ */
+
+package org.opensearch.dataprepper.plugins.kinesis.source.processor;
+
+import software.amazon.kinesis.processor.RecordProcessorCheckpointer;
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class KinesisCheckpointerTracker {
+    private final Map<ExtendedSequenceNumber, KinesisCheckpointerRecord> checkpointerRecordList = new LinkedHashMap<>();
+
+    public synchronized void addRecordForCheckpoint(final ExtendedSequenceNumber extendedSequenceNumber,
+                                                    final RecordProcessorCheckpointer checkpointer) {
+        checkpointerRecordList.put(extendedSequenceNumber, KinesisCheckpointerRecord.builder()
+                .extendedSequenceNumber(extendedSequenceNumber)
+                .checkpointer(checkpointer)
+                .readyToCheckpoint(false)
+                .build());
+    }
+
+    public synchronized void markSequenceNumberForCheckpoint(final ExtendedSequenceNumber extendedSequenceNumber) {
+        if (!checkpointerRecordList.containsKey(extendedSequenceNumber)) {
+            throw new IllegalArgumentException("checkpointer not available");
+        }
+        checkpointerRecordList.get(extendedSequenceNumber).setReadyToCheckpoint(true);
+    }
+
+    public synchronized Optional<KinesisCheckpointerRecord> getLatestAvailableCheckpointRecord() {
```

Comment:
Maybe rename this to `popLatestAvailableCheckpointRecord` or something similar. It is mutating the structure and a `get` method doesn't convey that.

Reply:
I have renamed this method to `popLatestReadyToCheckpointRecord`